pallene/constant_propagation.lua

```diff
@@ -16,128 +16,184 @@ local function is_constant_value(v)
     elseif tag == "ir.Value.Float"    then return true
     elseif tag == "ir.Value.String"   then return true
     elseif tag == "ir.Value.LocalVar" then return false
-    elseif tag == "ir.Value.Function" then return true
+    elseif tag == "ir.Value.Upvalue"  then return false
     else
         typedecl.tag_error(tag)
     end
 end
 
+-- Stores the constant initializers for local variables and upvalues of a function.
+local function FuncData(func)
+    local fdata = {
+        n_writes_of_locvar      = {}, -- { loc_id => integer  }
+        locvar_constant_init    = {}, -- { loc_id => ir.Value }
+        constant_val_of_upvalue = {}, -- { upv_id => ir.Value }
+        new_upvalue_id          = {}, -- { upv_id => upv_id   }
+    }
+
+    for loc_id = 1, #func.vars do
+        fdata.locvar_constant_init[loc_id] = false
+        fdata.n_writes_of_locvar[loc_id] = 0
+    end
+
+    for u_id = 1, #func.captured_vars do
+        fdata.constant_val_of_upvalue[u_id] = false
+    end
+
+    return fdata
+end
+
 -- Replaces toplevel constant variables by their respective values.
 --
 -- Currently assumes that the toplevel constant variable is initialized with a constant literal.
 -- Does not currently recognize non-trivial constant expressions as being constant.
 function constant_propagation.run(module)
 
-    local n_globals = #module.globals
-
-    -- 1) Find what toplevel variables are initialized to a constant in $init
+    -- 1) Find what toplevel variables are initialized to a constant.
 
-    local constant_initializer = {} -- { g_id => ir.Value? }
-    for i = 1, n_globals do
-        constant_initializer[i] = false
+    local data_of_func = {} -- list of FuncData
+    for _, func in ipairs(module.functions) do
+        table.insert(data_of_func, FuncData(func))
     end
 
-    do
-        -- DFS traversal to find SetGlobal instructions with a constant initializer. We ignore the
-        -- instructions inside If, Loop, and For statements, since those might be skipped.
-        local stack = { module.functions[1].body }
+
+    for f_id, func in ipairs(module.functions) do
+        -- DFS traversal to find the ir.Cmd.Move instructions which have constant
+        -- values as their src. We skip initializers in loops to make sure we only count
+        -- the initializers which are guaranteed to be evaluated at runtime.
+        local f_data   = assert(data_of_func[f_id])
+
+        local stack =  { func.body }
```

Comment:
Now that we are no longer avoiding if statements, would it be possible to use an `ir.iter` here?